docs/design/feature-attributes.md

```diff
@@ -0,0 +1,181 @@
+
+There is also the possibility that we will create an attribute-based model which allows users to define their own such attributes; see this draft for example: https://github.com/dotnet/designs/pull/261. The semantics outlined here could be extended to those attributes if we determine that they are appropriate there.
+
+## Goals
+
+- Define the semantics of feature attributes
+- Define the access patterns which are allowed and disallowed by these semantics
+
+## Non-goals
+
+- Specify the warning codes or wording of the specific warnings for disallowed access
+- Define a model for defining new feature attributes
+- Define an attribute-based model for feature switches
+- Define the interactions between `RequiresUnreferencedCodeAttribute` and `DynamicallyAccessedMembersAttribute`
+
+## RequiresFeatureAttribute
+
+`RequiresFeature` may be used on methods, constructors, or classes only.
+
+The use of this attribute establishes a [_feature requirement_](#feature-requirement) for the attributed type or member, which restricts access to the attributed type or member (and in some cases to other related IL) in certain ways. It also establishes a [_feature available_](#feature-available-scope) scope (which includes the attributed member but may also include other related IL) wherein access to members with a _feature requirement_ is allowed.
+
+Access to members with a _feature requirement_ is always allowed from a _feature available_ scope, and never produces feature warnings. The restrictions created by _feature requirement_ only limit access from scopes outside of _feature available_, where certain access patterns produce warnings.
+
+## Feature available scope
+
+Methods and constructors (including static constructors) with a _feature requirement_ are in a _feature available_ scope.
+
+Methods, constructors (including static constructors), and fields declared in a class or struct with a _feature requirement_ are also in a _feature available_ scope.
+
+Properties and events declared in a class or struct with a _feature requirement_ are also in a _feature available_ scope.
+
+Lambdas and local functions inside of a method in a _feature available_ scope are also in a _feature available_ scope.
+
+Note that nested types declared in a type that is in a _feature available_ scope are not necessarily in a _feature available_ scope.
+
+## Feature requirement
+
+### Methods
+
+When `RequiresFeature` is used on a method or constructor (except static constructors), this declares a _feature requirement_ for the method.
+
+Static constructors never have a _feature requirement_. `RequiresFeature` on a static constructor is not supported.
+
+### Classes
+
+When `RequiresFeature` is used on a class, this declares a _feature requirement_ for the class.
+
+When a class has a _feature requirement_, this creates a _feature requirement_ for the following members of the class:
```

Comment:
I think it's worth mentioning that this list IS the definition of "feature requirement on the class", there's no more to it. Specifically - accessing the type itself does not create a feature requirement - that is `typeof(Foo)` does not warn if `Foo` has a `RUC` on it.

Reply:
Also maybe of note that it makes no requirements/effect on the base class and interfaces.